### PR TITLE
Enable build for aarch64 in Dockerfile

### PR DESCRIPTION
**Pull request by andkrause**
This PR contains an "enhanced" version of the original Dockerfile that enables you to build cross-platform images for both amd64 and aarch64 with buildx. Sample command is below.

`docker buildx build --push --platform linux/arm64/v8,linux/amd64 --tag jbarlow83/ocrmypdf:latest --file .docker/Dockerfile .`

I will also try to work on a matching action to automate build upon merge.

### CODE REVIEW

**Comment by jbarlow83** [comment]
A few questions:

Is the issue just that those dependencies are not explicit?

I currently have Docker Hub set up to auto-build ocrmypdf whenever there's a new commit to master or tagged version. How would that work in this brave new world that acknowledges architectures other than x64? Is the idea that we'd have a Github Action or Azure Pipeline build step to generate and push the images, or can Docker Hub do that?

Does that buildx command generate two images, one for arm64 and one for amd64?

**Comment by andkrause** [comment]
The underlying issue is that precompiled wheels / dependencies are not available for aarch64. Hence I had to add several dependencies (mostly based on other issues associated to this repo) to allow pip to compile the respective modules. Now the Dockerfile can be used for docker build on both amd64 as well as aarch64 architectures. 

This also enables the outlined multi-architecture build (with buildx). In essence this means that 2 images will be built under the same tag. The docker daemon knows its underlying cpu architecture and will then pull the image matching its architecture. The details are really well described [here](https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/). I have used it to locally build this [clone](https://hub.docker.com/repository/docker/andy008/ocrmypdf/tags?page=1) of ocrmypdf with docker desktop for mac.

A real setup would of course use a CI/CD approach. Both Docker Hub and Github Actions seem to support this. I was going to explore github actions and could contribute a workflow later on. 

Hope that covers your questions. Please share your thoughts.

**Comment by jbarlow83** [comment]
I see, thanks for the explanation.

I'm currently using Azure Pipelines but Github Actions would be fine, since I will probably migrate everything there. Better to have one CI do it all in one shot.

**Comment by andkrause** [comment]
@jbarlow83: Github Workflow to execute buildx is now included. To use it, the following Github secrets need to be maintained:

- DOCKERHUB_IMAGE_NAME = ocrmypdf
- DOCKERHUB_USERNAME = jbarlow83
- DOCKERHUB_PASSWORD = Docker hub password (or [Token](https://docs.docker.com/docker-hub/access-tokens/)) 

Workflow is triggered upon push to master branch (latest tag) or release creation (release-name tag). It takes around 2 hours to complete.

Let me know if you have further questions.

**Comment by jbarlow83** [comment]
I will accept this, just hesitating on the "2 hours to complete" aspect. That's... lengthy.

**Comment by andkrause** [comment]
I was surprised as well, I basically see 2 reasons for it:

* Github does not really provide a lot of CPU (on my Macbook the entire thing takes 15 minutes)
* A lot of the requirements are not available as binaries for aarch64 and hence they are compiled at runtime. As my python is non-existant I have no idea how to fix it.

Cheers

Andreas

**Comment by QuLogic** [comment]
Are you building wheels, or a container? Because these dependencies may not be available as pre-compiled wheels for those architectures, but they may already be compiled in the distro (e.g., Fedora has all of your deps for aarch64, and I'm sure Debian does too.)

**Comment by jbarlow83** [comment]
This is a container build, but as currently written the Dockerfile uses PyPI wheels to satisfy many of the dependencies, rather than the operating system's.

If someone just want container isolation of an OS version, they can write a trivial Dockerfile to `dnf install ocrmypdf`.

If someone upgrade ocrmypdf beyond the OS version, using the OS version of everything else, `pip install --user ocrmypdf` should cover them.

So the use case of the ocrmypdf Docker image is for people who want more recent versions of dependencies than their OS may provide (or jbig2enc which is not distributed because of its lingering patent fears), which I think justifies using Python wheels as we do in the Docker image.

Definitely the best way to get a non-x86 ocrmypdf right now is to use a pre-compiled distro.

**Comment by andkrause** [comment]
@jbarlow83: is there any hope this will be merged any time soon?

**Comment by jbarlow83** [comment]
The project license has changed since you submitted this PR. Are you okay with the new license?

**Comment by andkrause** [comment]
works for me

File: .docker/Dockerfile [hunk 1]
```diff
@@ -11,10 +11,19 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
+  zlib1g-dev \
+  libqpdf-dev \
+  pybind11-dev \
+  libxml2-dev \ 
+  libxslt1-dev \
   python3-distutils \
   ca-certificates \
   curl \
-  git
+  git
```
Comment (0x326) [inline .docker/Dockerfile:26]: There shouldn't be a trailing white space character here
```suggestion
  git
```

File: .docker/Dockerfile [hunk 2]
```diff
@@ -11,10 +11,19 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
   python3 \
+  python3-dev \
+  libssl-dev \
+  libffi-dev \
+  libjpeg-dev \
+  zlib1g-dev \
+  libqpdf-dev \
+  pybind11-dev \
+  libxml2-dev \ 
+  libxslt1-dev \
```
Comment (0x326) [inline .docker/Dockerfile:22]: When I added `aarch64` support on my own, I only needed:
```
  libffi-dev \
  libqpdf-dev \
  python3-dev \
```
However, adding more probably won't hurt

File: .github/workflows/buildx-latest.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        run: echo ${{ secrets.DOCKERHUB_PASSWORD }} | docker login -u ${{ secrets.DOCKERHUB_USERNAME }} --password-stdin 2>&1
+      - name: Print Image Tag
+        run: echo "Building Image ${DOCKER_REPOSITORY}/${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG}"
+      - name: Run Buildx
+        run: |
+          docker buildx build \
+            --push \
+            --platform linux/arm64/v8,linux/amd64  \
+            --tag "${DOCKER_REPOSITORY}/${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG}" \
+            --file .docker/Dockerfile .
```
Comment (0x326) [inline .github/workflows/buildx-latest.yml:36]: Missing ending newline

File: .docker/Dockerfile [hunk 4]
```diff
@@ -52,6 +61,7 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
   liblept5 \
   libsm6 libxext6 libxrender-dev \
   zlib1g \
+  libxslt-dev \
```
Comment (0x326) [inline .docker/Dockerfile:64]: Is this needed? I didn't need it when I built `v11.0.2`

File: .docker/Dockerfile [hunk 5]
```diff
@@ -62,7 +72,8 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
   tesseract-ocr-fra \
   tesseract-ocr-por \
   tesseract-ocr-spa \
-  unpaper
+  unpaper \
+  && rm -rf /var/lib/apt/lists/*
```
Comment (0x326) [inline .docker/Dockerfile:76]: Nice addition, but unrelated to `aarch64` support

**Review by 0x326 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**